crates/notedeck/src/submgr.rs

```diff
@@ -0,0 +1,269 @@
+}
+
+pub type SubResult<T> = Result<T, SubError>;
+
+#[derive(Debug, Clone, Copy)]
+pub struct SubId(nostrdb::Subscription);
+
+impl From<Subscription> for SubId {
+    fn from(subscription: Subscription) -> Self {
+        SubId(subscription)
+    }
+}
+
+impl Ord for SubId {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.0.id().cmp(&other.0.id())
+    }
+}
+
+impl PartialOrd for SubId {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for SubId {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.id() == other.0.id()
+    }
+}
+
+impl Eq for SubId {}
+
+#[derive(Debug, Clone)]
+pub enum SubConstraint {
+    OneShot,                    // terminate subscription after initial query
+    Local,                      // only query the local db, no remote subs
+    OutboxRelays(Vec<String>),  // ensure one of these is in the active relay set
+    AllowedRelays(Vec<String>), // if not empty, only use these relays
+    BlockedRelays(Vec<String>), // if not empty, don't use these relays
+}
+
+#[derive(Debug, Default)]
+pub struct SubSpecBuilder {
+    rmtid: Option<String>,
+    filters: Vec<Filter>,
+    constraints: Vec<SubConstraint>,
+}
+
+impl SubSpecBuilder {
+    pub fn new() -> Self {
+        SubSpecBuilder::default()
+    }
+    pub fn rmtid(mut self, id: String) -> Self {
+        self.rmtid = Some(id);
+        self
+    }
+    pub fn filters(mut self, filters: Vec<Filter>) -> Self {
+        self.filters.extend(filters);
+        self
+    }
+    pub fn constraint(mut self, constraint: SubConstraint) -> Self {
+        self.constraints.push(constraint);
+        self
+    }
+    pub fn build(self) -> SubSpec {
+        let mut outbox_relays = Vec::new();
+        let mut allowed_relays = Vec::new();
+        let mut blocked_relays = Vec::new();
+        let mut is_oneshot = false;
+        let mut is_local = false;
+
+        for constraint in self.constraints {
+            match constraint {
+                SubConstraint::OneShot => is_oneshot = true,
+                SubConstraint::Local => is_local = true,
+                SubConstraint::OutboxRelays(relays) => outbox_relays.extend(relays),
+                SubConstraint::AllowedRelays(relays) => allowed_relays.extend(relays),
+                SubConstraint::BlockedRelays(relays) => blocked_relays.extend(relays),
+            }
+        }
+
+        SubSpec {
+            rmtid: self.rmtid,
+            filters: self.filters,
+            outbox_relays,
+            allowed_relays,
+            blocked_relays,
+            is_oneshot,
+            is_local,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct SubSpec {
+    rmtid: Option<String>,
+    filters: Vec<Filter>,
+    outbox_relays: Vec<String>,
+    allowed_relays: Vec<String>,
+    blocked_relays: Vec<String>,
+    is_oneshot: bool,
+    is_local: bool,
+}
+
+pub struct SubMgr<'a> {
+    ndb: &'a mut Ndb,
```

Comment:
You can just use Ndb, they are cleap to clone and can send them between threads (they are atomically reference counted). holding mutable references will probably be a pain. SubscriptionStream's hold Ndb clones internally as well.

Reply:
Done,  fixed the issues